main/src/ca/uwaterloo/flix/language/phase/Verifier.scala

```diff
@@ -207,257 +211,287 @@ object Verifier {
+              )
             case _ => failMismatchedShape(t1, "Array", loc)
           }
 
         case AtomicOp.Ref =>
-          val List(t1) = ts
+          val (List(t1), List(p1)) = rs.unzip
           val refType = MonoType.Ref(t1)
-          checkEq(refType, tpe, loc)
+          (checkEq(refType, tpe, loc), checkEq(pur, Purity.combine(pur, p1), loc))
 
         case AtomicOp.Deref =>
-          val List(t1) = ts
+          val (List(t1), List(p1)) = rs.unzip
           t1 match {
-            case MonoType.Ref(elm) => checkEq(elm, tpe, loc)
+            case MonoType.Ref(elm) =>
+              (checkEq(elm, tpe, loc), checkEq(pur, Purity.combine(pur, p1), loc))
             case _ => failMismatchedShape(t1, "Ref", loc)
           }
 
         case AtomicOp.Lazy =>
-          val List(t1) = ts
+          val (List(t1), List(p1)) = rs.unzip
           tpe match {
             case MonoType.Lazy(elmt) =>
               val fun = MonoType.Arrow(List(MonoType.Unit), elmt)
               checkEq(t1, fun, loc)
-              tpe
+              check(expected = Purity.Pure)(actual = p1, loc)
+              (tpe, check(expected = Purity.Pure)(actual = pur, loc))
             case _ => failMismatchedShape(tpe, "Lazy", loc)
           }
 
         case AtomicOp.Force =>
-          val List(t1) = ts
+          val (List(t1), List(p1)) = rs.unzip
           t1 match {
-            case MonoType.Lazy(elm) => checkEq(elm, tpe, loc)
+            case MonoType.Lazy(elm) =>
+              check(expected = Purity.Pure)(actual = p1, loc)
+              (checkEq(elm, tpe, loc), check(expected = Purity.Pure)(actual = pur, loc))
             case _ => failMismatchedShape(t1, "Lazy", loc)
           }
 
         case AtomicOp.Tuple =>
+          val (ts, ps) = rs.unzip
           val tup = MonoType.Tuple(ts)
-          checkEq(tup, tpe, loc)
+          (checkEq(tup, tpe, loc), checkEq(pur, Purity.combineAll(ps), loc))
 
         case AtomicOp.Index(idx: Int) =>
-          val List(t1) = ts
+          val (List(t1), List(p1)) = rs.unzip
           t1 match {
-            case MonoType.Tuple(elms) => checkEq(elms(idx), tpe, loc)
+            case MonoType.Tuple(elms) =>
+              (checkEq(elms(idx), tpe, loc), checkEq(pur, p1, loc))
             case _ => failMismatchedShape(t1, "Tuple", loc)
           }
 
         case AtomicOp.Assign =>
-          val List(t1, t2) = ts
+          val (List(t1, t2), List(p1, p2)) = rs.unzip
           t1 match {
             case MonoType.Ref(elm) =>
               checkEq(t2, elm, loc)
-              check(expected = MonoType.Unit)(actual = tpe, loc)
+             (check(expected = MonoType.Unit)(actual = tpe, loc), checkEq(pur, Purity.combine3(pur, p1, p2), loc))
             case _ => failMismatchedShape(t1, "Ref", loc)
           }
 
         // Match- and Hole-errors match with any type
         case AtomicOp.HoleError(_) =>
-          tpe
+          (tpe, pur)
```

Comment:
i might be mistaken, but i think these error should always be pure

Reply:
i ran into some cases where Hole/MatchError's were impure.